scripts/front-end/actions/main.js

```diff
@@ -14,15 +14,18 @@ export function chargerDossiers(){
     console.log('store.state.secret', store.state.secret)
     if(store.state.secret){
         return json(`/dossiers?secret=${store.state.secret}`)
-            .then(/** @type {Dossier[]} */ dossiers => {
```

Comment:
Ici, ça ne marchait pas du tout le typage du paramètres `dossiers`. Ça ne matchait pas avec la signature de la méthode `json`.

Reply:
yes
`json` retourne genre un `Promise<unknown>` et c'est "bien" parce qu'effectivement, la fonction générique de téléchargement `json` ne sait pas ce qu'elle retourne comme objet, donc c'est cool qu'elle ne se mouille pas... et en même temps, ça l'empêche de nous aider ensuite

la manière propre que j'ai trouvé de gérer les `unknown`, c'est de créer un [type guard](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) pour faire du "narrowing"

ça marche genre comme ça : 

```js
/**
* un type-guard manuel
* 
* @params {any} x
* @returns {x is Dossier[]} // une attention particulière à cette ligne
*/
function isDossierArray(x){
  return Array.isArray(x) && x.every(isDossier) // isDossier retourne un type {y is Dossier} ; à écrire :-p
}

// ... ailleurs ....

json(url).then(dossiers => {
  if( ! isDossierArray(dossiers) ){
    throw new Error('On attendait des dossiers, sacrebleu !')
  }

  // ici, TypeScript sait que isDossierArray(dossiers) === true
  // et donc, TypeScript sait désormais que "dossiers" est de type Dossier[] et n'est plus unknown
})
```

Pendant longtemps, je trouvais ça balourd

et puis, suffisamment de fois, l'objet JSON récupéré a changé côté serveur et donc ça me créait une erreur côté front-end que je ne détectais que tard parce que c'était profondément dans un objet

et donc, j'ai commencé à comprendre la sagesse derrière `unknown` et l'intérêt du narrowing et des type-guard custom. Ça permet d'assurer une continuité garantie sur les types entre le serveur et le client

Plusieurs trucs sur les type-guard : 
- TypeScript n'est pas capable de comprendre si le code à l'intérieur du type-guard permet bien de vérifier que `x is Dossier[]`, on pourrait mettre `return Math.random() > 0.5` qu'il n'y verrait que du feu. Il fait confiance qu'après un appel à `isDossierArray(x)`, `x` est bien de type `Dossier[]` (et donc il offre la complétion et les vérifications en conséquent) uniquement grâce à la signature de la fonction
- Donc, c'est possible d'écrire un type-guard naïf/incomplet et d'y revenir plus tard
- Si la vérification de type est lourde niveau perf (genre vérifier qu'un millier d'objets ont une centaine de propriétés complexes chacun, tmtc), c'est possible d'avoir une version exhaustive en dev et une version légère (`return true` à l'extrême) en prod
- Pour l'avoir essayé, c'est aussi possible de donner la définition de l'interface `Dossier` à un LLM et lui demander d'écrire une fonction qui retourne un `x is Dossier[]`. Une fois que tu as le type, l'écriture d'une première version de type-guard du type peut être assez mécanique

Reply:
Si c'est ok pour toi et si l'exercice t'intéresse, est-ce que tu serais ok de faire un type-guard pour cette ligne-là ? (et vérifier que la variable `dossiers` sort bien de `unknown`)

Reply:
Je me suis servie d'un LLM bien connu :D et j'ai corrigé 2-3 trucs à la main. J'ai mis ça dans ce fichier : [scripts/types/typeguards.js](https://github.com/betagouv/pitchou/pull/66/files#diff-fdbefa7aa3dfe0d9808766ac722069446ec829de72724e6e27ad6668f4795f5b)